client/app/pages/queries/hooks/useQueryDataSources.js

```diff
@@ -0,0 +1,29 @@
+import { filter, find } from "lodash";
+import { useState, useMemo, useEffect } from "react";
+import { DataSource } from "@/services/data-source";
+
+export default function useQueryDataSources(query) {
+  const [allDataSources, setAllDataSources] = useState([]);
+  const [dataSourcesLoaded, setDataSourcesLoaded] = useState(false);
+  const dataSources = useMemo(() => filter(allDataSources, ds => !ds.view_only || ds.id === query.data_source_id), [
+    allDataSources,
+    query.data_source_id,
+  ]);
+  const dataSource = useMemo(() => find(dataSources, { id: query.data_source_id }) || null, [query, dataSources]);
+
+  useEffect(() => {
+    let cancelDataSourceLoading = false;
+    DataSource.query().$promise.then(data => {
+      if (!cancelDataSourceLoading) {
+        setDataSourcesLoaded(true);
+        setAllDataSources(data);
+      }
+    });
+
+    return () => {
+      cancelDataSourceLoading = true;
+    };
+  }, []);
+
+  return useMemo(() => ({ dataSourcesLoaded, dataSources, dataSource }), [dataSourcesLoaded, dataSources, dataSource]);
```

Comment:
This `useMemo` was to avoid unnecessary updates when listing dependencies with the hook value, right?

Reply:
I think it's possible to use only the attributes in the dependencies without having to disable eslint:
```js
[hook.value] /* instead of */ [hook]
```

LMK you already tried this :)

Even with the `useMemo`, it makes sense since it would avoid unnecessary updated when a non-related value changes.